src/nix/weather.cc

```diff
@@ -0,0 +1,191 @@
+                for (auto & [cursor, prefix] : installable2->getCursors(*state)) {
+                    visit(*cursor, parseAttrPath(*state, prefix), true);
+                }
+            } else {
+                auto drvPaths_ = toDerivations(store, {installable}, true);
+                drvPaths.insert(drvPaths_.begin(), drvPaths_.end());
+            }
+        }
+
+        if (drvPaths.size() == 0)
+            throw Error("no derivations found!");
+
+        StorePathSet closure;
+        if (!noClosure) {
+            Activity act(*logger, lvlInfo, actUnknown, "computing closure");
+            store->computeFSClosure(std::move(drvPaths), closure); // this could be huge ; should we chunk it up?
+        } else
+            closure = std::move(drvPaths);
+
+        StorePathSet outPaths;
+        for (auto & path : closure) {
+            Activity act(*logger, lvlInfo, actUnknown, fmt("querying '%s' output paths", store->printStorePath(path)));
+            if (isDerivation(store->printStorePath(path))) {
+                auto outputs = store->queryDerivationOutputs(path);
+                outPaths.insert(outputs.begin(), outputs.end());
+            }
+            // TODO: what to do with non-derivations?
+        }
+
+        StorePathSet totalValidPaths;
+        ssize_t totalPathsFound = 0;
+        uint64_t totalNarSize = 0;
+        std::optional<uint64_t> totalDownloadSize = 0;
+
+        for (auto & sub : subs) {
+            StorePathSet validPaths = sub->queryValidPaths(outPaths);
+
+            ssize_t pathsFound = 0;
+            uint64_t narSize = 0;
+            std::optional<uint64_t> downloadSize = 0;
+
+            for (auto & path : validPaths) {
+                auto info = sub->queryPathInfo(path);
+                auto narInfo = std::dynamic_pointer_cast<const NarInfo>(std::shared_ptr<const ValidPathInfo>(info));
+
+                pathsFound++;
+                if (downloadSize && narInfo)
+                    *downloadSize += narInfo->fileSize;
+                else
+                    downloadSize = std::nullopt;
+                narSize += info->narSize;
+
+                if (totalValidPaths.count(path) == 0) {
+                    totalValidPaths.insert(path);
+                    totalPathsFound++;
+                    if (totalDownloadSize && narInfo)
+                        *totalDownloadSize += narInfo->fileSize;
+                    else
+                        totalDownloadSize = std::nullopt;
+                    totalNarSize += info->narSize;
+                }
+            }
+
+            logger->cout("Substituter %s", sub->getUri());
+            logger->cout("  %6.1f%% of paths have substitutes available (%s of %s)", (float) 100 * pathsFound / outPaths.size(), pathsFound, outPaths.size());
+            if (downloadSize)
+                logger->cout("  %s downloaded (compressed)", formatSize(*downloadSize));
+            logger->cout("  %s downloaded (uncompressed)", formatSize(narSize));
```

Comment:
"downloaded (compressed)" and `"downloaded` (uncompressed)" is confusing to me. Maybe just "compressed size" (or "download `size)` and "uncompressed size"?